Fix initialization race-condition in runtime connector

hubertp opened the pull request:
### Pull Request Description

It seems that Runtime Connector wasn't respecting the protocol it defined itself. The connector should be waiting on the `Api.InitializedNotification` message and only then start forwarding messages. So far it seems this hasn't been a problem, or at least wasn't reported as such, because initialization was fast enough.

Modified `Handler` so that we are certain that its fields hold initialized values when being accessed by different threads.

Should fix problems mentioned in #7898.

### Important notest

Fixing the initial problem revealed some occasional deadlocks that can happen between command execution processor and context locks. Had to partially revert a fix in #6998 by making sure that all commands are executed in a non-blocking way while still preserving the order for some of them.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests continue to pass.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

hubertp committed:
Fix initialization race-condition in runtime connector

It seems that Runtime Connector wasn't respecting the protocol it
defined itself. The connector should be waiting on the
`Api.InitializedNotification` message and only then start forwarding
messages. So far it seems this hasn't been a problem, or at least wasn't
reported as such, because initialization was fast enough.

Modified `Handler` so that we are certain that its fields hold
initialized values when being accessed by different threads.

4e6 commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/Handler.scala:
```diff
@@ -116,15 +119,17 @@ final class Handler {
     case request: Api.Request =>
-      val cmd = CommandFactory.createCommand(request)
-      commandProcessor.invoke(cmd)
+      if (ctx != null) {
```
The case when the context is not initialized, and the handler already started receiving messages most likely indicates an error somewhere else. We should at least log it appropriately.

hubertp replied:
Yeah, except without having a reference to `TruffleContext` I don't have the logger either. Need to figure out how to pass it around.

hubertp replied:
Throwing an exception. This is an internal error that should always be satisfied

hubertp commented:
Converted to draft since I discovered some problems.

hubertp committed:
forgot to unstash messages

hubertp committed:
Execute synchronous commands via separate executor

Rather than using an overly complicated locking mechanism switched to a
`BlockingQueue` and a separate thread pool quaranteeing that commands
are executed in order but they are also non-blocking.

This change partially reverts #6998 which introduced synchronous
commands to a) avoid deadlocks b) ensure the correct execution of
close/open commands c) avoid data corruption when applying edits.

Sadly, it now turns out that the possibility of synchronous commands is
at odds with how commands are expected to execute. More importantly,
there are commands comming from Language Server but there are commands
comming from Runtime as well. When combined with context locks it
creates an environment for deadlocks as described in #7898.

hubertp committed:
Throw an exception when handler isn't ready

hubertp commented:
Deadlocks no longer appear as I've eliminated one element - synchronous commands. Now compilation, and a potential (locking) command when ensuring library is loaded, will not block.

hubertp committed:
Don't wait infinitely for an answer

Lock acquisition should finish almost instantly. If it doesn't, then
something is wrong (spoiler alert - it really was).

hubertp commented:
Once this is fixed, there are still problems in suggestions but they appear once edits are being attempted. I will file a separate ticket for that.

radeusgd commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ReentrantLocking.scala:
```diff
@@ -126,4 +132,57 @@ class ReentrantLocking(logger: TruffleLogger) extends Locking {
+  }
+
+  private class OrderedTaskIndex {
+    private var current = 0
+    private var max     = 0
```
I'd give these explicit types, to be clearer what they are.

radeusgd commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ReentrantLocking.scala:
```diff
@@ -126,4 +132,57 @@ class ReentrantLocking(logger: TruffleLogger) extends Locking {
+      current
+    }
+
+    def enqueue(): Int = {
+      val idx = max
+      max = max + 1
+      assert(max > current)
+      idx
+    }
```
What happens if the index overflows? IMO assertion is not enough - we should have a check that runs without asserts enabled as well.

We may also change this to `Long` to give it more space before it overflows

radeusgd commented on lib/scala/connected-lock-manager/src/main/scala/org/enso/lockmanager/client/ConnectedLockManager.scala:
```diff
@@ -81,17 +83,17 @@ class ConnectedLockManager extends LockManager {
 
   private def sendRequestAndWaitForResponse(
     request: Runtime.ApiRequest
-  ): Runtime.ApiResponse = {
+  ): Try[Runtime.ApiResponse] = {
     val future = getEndpoint.sendRequest(request)
-    Await.result(future, Duration.Inf)
+    Try(Await.result(future, Duration(3, TimeUnit.SECONDS)))
   }
```
This timeout is far far far too small.

I remember my confusion why I put the infinite timeout here but now I remember and it did have its merit after all.

The reason is - we are here trying to acquire a lock for library resources. If another process is in the middle of installing that library, the lock will need to wait the whole time until the library finishes installation. With bad internet connection, for some big libraries that we may have in the future, this may even take _hours_.

Well, I think we still need a timeout - the compiler getting stuck for hours because of a background job doing installation is no good thing either - if the installation is taking longer than expected - fine - we just need to cancel the current job. We just should tell the user clearly what happened.

So I suggest:
1. Let's make the timeout a bit larger, e.g. 30 or 60 seconds. This way a routine installation happening in the background will not crash the compiler but just make it wait a bit. I guess in many cases an installation _will_ take longer than 10s, so any timeout less than that is IMO insufficient.
2. I think ideally we should detect the `TimeoutException` in `ResourceManager` and there convert it into a more user friendly message - e.g. "Timed out when waiting to acquire a lock for library Foo.Bar, perhaps it is being modified in the background. Please wait until all background installation finish and restart the engine to try again.".

hubertp replied:
I increased the timeout but I don't like the current API either. `withResource` and `withResources` should return a `Try` rather than throwing undeclared exceptions. Changing that will however trigger an avalanche of other changes that don't belong to this PR and should be addressed separately.

radeusgd commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/command/PartiallyOrderedAsynchronousCommand.scala:
```diff
@@ -0,0 +1,91 @@
+    ec: ExecutionContext,
+    idx: Int
+  ): Future[Unit] = {
+    ctx.locking.waitOnTaskIndex(idx)
+    try {
+      executeAsynchronously
+    } finally {
+      ctx.locking.advanceTaskIndex()
+    }
```
Is this surely correct?

Inside of this `try` block, we run `executeAsynchronously` which prepares, starts and returns a `Future[Unit]`. We then enter the `finally` block and advance the task index, but we don't know if the `Future` that we are about to return has finished execution.


---

Do we have some tests that check the synchronization of this? e.g. for OpenFileCmd etc.

I think this should be tested. It's not a trivial piece of logic. We have infrastructure for testing parallelism and synchronization, see for example `TestSynchronizer` and its usages.

radeusgd commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ReentrantLocking.scala:
```diff
@@ -126,4 +132,57 @@ class ReentrantLocking(logger: TruffleLogger) extends Locking {
+    }
+  }
+
+  /** @inheritdoc */
+  override def waitOnTaskIndex(index: Int): Unit = {
+    orderLock.lock()
+    try {
+      val current = taskIndex.getCurrent()
+      while (current != index) {
+        indexAdvancedCondition.await()
+      }
+    } finally {
+      orderLock.unlock()
+    }
+  }
```
How do we handle the possibility of `InterruptedException` here?

radeusgd replied:
How do we guarantee the liveness of the `PartiallyOrderedAsynchronousCommand` queue?

What happens if I have 4 threads with task ids 1, 2, 3 and 4. 1 is executing a long running operation, thus 2, 3 and 4 are waiting.

In the meantime, something kills the thread associated with (3). Once 1 finishes execution, 2 starts and then finishes. Now the `indexAdvancedCondtion` is signalled. But since (3) is dead, only (4) wakes up - it sees that its not yet its time - the current index is set at 3, and goes to sleep again. But now it will never wake up, because 3 will never advance the clock.

radeusgd commented:
I think the idea of `PartiallyOrderedAsynchronousCommand` is perfect for this issue and it is very elegant.

However, I'm not absolutely certain that the current implementation guarantees correctness (IMO the `advanceTaskIndex` in `finally` happens too early) nor liveness (if a thread waiting on the queue gets interrupted, IMO all subsequent tasks will never get executed).

Could we maybe simplify the whole logic? Instead of using the signals, wouldn't it be easier to put new `PartiallyOrderedAsynchronousCommand` onto a `ConcurrentQueue` and have one executor thread that will `pop` them from the queue, execute them synchronously (either without `Future`, or if with `Future`, it will wait until the future completes with a success or error) and then picks next tasks? IMO that architecture would be simpler to understand and will guarantee both correctness and liveness.

JaroslavTulach commented on CHANGELOG.md:
```diff
@@ -979,6 +979,7 @@
 - [Add Enso-specific assertions][7883]
 - [Modules can be `private`][7840]
 - [Export of non-existing symbols results in error][7960]
+- [Improve backend initialization when more clients are present][7985]
```
_CI: No change log needed_ may also be useful in case of small bugfixes.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/command/SerializeModuleCommand.java:
```diff
@@ -20,8 +20,8 @@ public SerializeModuleCommand(Option<UUID> maybeRequestId, QualifiedName moduleN
 
   @Override
-  public Future<BoxedUnit> execute(RuntimeContext ctx, ExecutionContext ec) {
+  public Future<Completion> execute(RuntimeContext ctx, ExecutionContext ec) {
```
I have no idea what's the difference of returning `Future<Unit>` and `Future<Completion>`.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/Handler.scala:
```diff
@@ -116,15 +119,21 @@ final class Handler {
     case Api.Request(requestId, Api.ShutDownRuntimeServer()) =>
-      if (commandProcessor ne null) {
-        commandProcessor.stop()
+      if (ctx != null) {
```
Reading into local variable first and then checking and using is safer/faster. Two accesses to `ctx` volatile field will need two reads from the global memory.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/command/AsynchronousCommand.scala:
```diff
@@ -13,12 +13,11 @@ import scala.util.{Failure, Success}
     ctx: RuntimeContext,
     ec: ExecutionContext
-  ): Result[Completion] = {
+  ): Future[Completion] = {
```
What is the difference between `Result` and `Future` in Akka?

hubertp replied:
`Result` is a type alias defined in `Command`. For `AsynchronousCommand` it is `Result[T] = Future[T]` but for `SynchronousCommand` it is `Result[T] = T`. A small trick from Scala to make compiler happy and still have some typesafety.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/command/PartiallyOrderedAsynchronousCommand.scala:
```diff
@@ -0,0 +1,91 @@
+    } yield result
+  }
+
+  private def mapFailures(logger: TruffleLogger, result: Future[Unit])(implicit
```
Why don't this `mapFailures` method doesn't accept `Future[Completion]` but returns one?

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/command/PartiallyOrderedAsynchronousCommand.scala:
```diff
@@ -0,0 +1,91 @@
+    ec: ExecutionContext,
+    idx: Int
+  ): Future[Unit] = {
+    ctx.locking.waitOnTaskIndex(idx)
```
`ctx.locking.waitOnTaskIndex`!? Was the concept already present or is it a new idea?

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/Locking.scala:
```diff
@@ -75,4 +75,19 @@ trait Locking {
+  /** Returns a unique number, monotonically increasing, that can be used for ordering execution.
+    * @return integer value that identifies current execution state
+    */
+  def getUniqueTaskIndex(): Int
```
Index into something to wait upon!? What a _C-like_ concept. Why don't we use regular primitives like locks? For example:
```java
<T> Future<T> createNewTask(Consumer<T> resolve);
```
It'd be better to work with types `Future` objects and with _indexes_

JaroslavTulach commented on engine/runtime-instrument-id-execution/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java:
```diff
@@ -257,9 +258,9 @@ public EventBinding<ExecutionEventNodeFactory> bind(
-      final int afterFunctionLine = c.getSourceSection().getEndLine() + 1;
+    if (entryCallTarget instanceof RootCallTarget r && r.getRootNode() instanceof ClosureRootNode c && c.getSourceSection() instanceof SourceSection section && section != null) {
+      final int firstFunctionLine = section.getStartLine();
+      final int afterFunctionLine = section.getEndLine() + 1;
```
Safer. Thank you.

JaroslavTulach reviewed (approved):
I cannot comment much on the Scala code, except I don't think using _indexes_ is _object oriented enough_. Engine changes seem to be fine.

hubertp committed:
nits

hubertp commented:
@radeusgd @JaroslavTulach thanks for the suggestions. I realized the ordering was too low level and also unnecessary. It was much easier to solve the problem with a blocking queue and a separate single-threaded executor.

4e6 reviewed (approved):
A separate executor is definitely better

radeusgd commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/CommandExecutionEngine.scala:
```diff
@@ -59,8 +76,23 @@ class CommandExecutionEngine(interpreterContext: InterpreterContext)
-    cmd.execute(runtimeContext, commandExecutionContext)
+  def invoke(cmd: Command): Future[Completion] = {
+    cmd match {
+      case c: SynchronousCommand =>
+        blockingQueue.add(c)
+        val javaFuture =
+          sequentialExecutionService.submit(new Callable[Completion] {
+            override def call(): Completion = {
+              val orderedCommand = blockingQueue.take()
+              orderedCommand.execute(runtimeContext, sequentialExecutionContext)
+            }
+          })
+        Future {
+          javaFuture.get()
+        }
```
I like how much simpler this approach is and yet it guarantees the ordering.

hubertp committed:
increase timeout

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/CommandExecutionEngine.scala:
```diff
@@ -59,8 +76,23 @@ class CommandExecutionEngine(interpreterContext: InterpreterContext)
+            }
+          })
+        Future {
+          javaFuture.get()
```
Creating new future that will block a thread and wait for Java future!? We don't have _virtual threads_ yet and regular threads take significant system resources.

Why not `javaFuture.asScala` as https://stackoverflow.com/questions/35033290/convert-a-java-future-to-a-scala-future suggests?

hubertp replied:
`asScala` is an implicit on `CompletableFuture` not `Future` so that won't work out of the box. But let me tinker with it.

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/CommandExecutionEngine.scala:
```diff
@@ -59,8 +76,23 @@ class CommandExecutionEngine(interpreterContext: InterpreterContext)
   /** @inheritdoc */
-  def invoke(cmd: Command): cmd.Result[Completion] = {
-    cmd.execute(runtimeContext, commandExecutionContext)
+  def invoke(cmd: Command): Future[Completion] = {
```
Is it guaranteed the order of commands sent to `invoke(cmd)` method is the same as messages read from the web socket stream?

If not, then the `blockingQueue.add(c)` would still not guarantee the order...

hubertp replied:
It is, yes.

hubertp committed:
s/Future/CompletableFuture

hubertp committed:
PR review

hubertp commented:
Figured out why we have those randomly failing tests in this PR.
Consider a sequence of API requests that trigger `OpenFileCmd`, `OpenFileCmd` and `PushContextCmd` in https://github.com/enso-org/enso/blob/develop/engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala#L653. Except that's not how they can end up executing because `OpenFileCmd` sends no runtime reply and `PushContextCmd` can easily end up between the two open file commands. 
In real life with actual handler response that wouldn't happen because the client should wait before issuing the request to push the context.

hubertp committed:
Made push/pop context cmds synchronous

This avoids a potential race-condition that could happen between
open/close file cmds and push/pop context cmds, as illustrated by
randomly failing tests in `runtime-with-instruments`. It was also pretty
easy to demonstrate once a `Thread.sleep` was artificially added to
`OpenFileCmd`. That way we make sure that `PushContextCmd` **always**
follows after `OpenFileCmd` finishes.

radeusgd commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/command/SynchronousCommand.scala:
```diff
@@ -36,6 +44,9 @@ abstract class SynchronousCommand(maybeRequestId: Option[RequestId])
     }
   }
 
-  def executeSynchronously(implicit ctx: RuntimeContext): Unit
+  def executeSynchronously(implicit
+    ctx: RuntimeContext,
+    ec: ExecutionContext
+  ): Unit
```
Sorry for post-mortem comment, but I haven't seen this before.

---

Why is the argument `ec: ExecutionContext` added here??

If I see correctly - it is not used anywhere anyway.

And it seems rather dangerous - the computation performed by this overload is supposed to happen synchronously, so we should not need an `ExecutionContext` for it, right?

hubertp replied:
Because of `PushContextCmd`. It needs to be in the same execution queue as open/close commands because there is a chance that client will issue it *before* it has a chance to run, resulting in some obscure NPEs. This is not a hypothetical solution and was quite frequent in our tests on slower machines.

Having said that, I'm going to rewrite it a bit so that open/close commands actually issue replies. That way we can ensure the correct order of execution, and then drop `PushContextCmd` from synchronous commands.
